### ISSUE TITLE
Replace dependency resolution with MicroDi

### ISSUE THREAD

**Issue by craigfowler** [csf-dev/CSF.Screenplay#65]
The plan is to replace the DI container within Screenplay with [MicroDi](https://github.com/csf-dev/MicroDi), which is itself a fork of BoDi (my fork). BoDi was written for SpecFlow but I need some extra features of my own and adding them directly to BoDi seemed too complex without first forking it and making it a little more maintainable.

## Archived comments removed
A number of comments were made in this issue which have now been removed as they are no longer relevant. They pertained to decision-making as to which DI container is most suitable.

In short: Autofac and Ninject are considered 'too big' to use here. BoDi looks good but it's missing a few features I need in order to implement screenplay (see list of required features in comments below).

Thus, I have decided to fork BoDi and extend it. In doing so I have needed to refactor it, because its "single source file" design is somewhat difficult to comprehend and maintain.

**Comment by craigfowler** [comment]
Turns out that BoDi was written specifically for SpecFlow and has no features which SpecFlow doesn't need. So actually that's not quite what I want, because it has no concept of scope support.

**Comment by craigfowler** [comment]
As a fallback position, if no suitable container can be found, then perhaps its worth separating the container used in this project out into a separate project of its own?

**Comment by craigfowler** [comment]
I'm not seeing any off the shelf containers which quite do what I want here. I don't want to go fully into something like Autofac, but BoDi isn't quite good enough either.

Looks like it will be down to wrapping mine up as a separate project.

**Comment by craigfowler** [comment]
Another argument for using an off the shelf container is that I'm increasingly needing advanced capabilities.

**Comment by craigfowler** [comment]
Adding to milestone 0.8.

Whilst this is not important from a functionality/usefulness perspective, it's a high-risk change which would introduce a lot of change into the API. It would almost certainly not be backwards-compatible.

**Comment by craigfowler** [comment]
After a second look, it looks like the BoDi container does in fact support nesting (and thus lifetime scopes). So, this might be suitable for use as my container of choice for Screenplay.

**Comment by craigfowler** [comment]
If I do use BoDi, then I want to include it as a NuGet package.  I know that it can be included as a single source file, but using a package will help consolidate versions across implementations.

I do need to find out which minimum version I can rely on.

**Comment by craigfowler** [comment]
# Requirements for the Screenplay DI container
These are the things which the container must be able to do/provide.

## Named registrations
A registration is given a name so we can differentiate it from other registrations of the same type.

## Typed registrations
This is the default registration. Also, typed registrations must be able to register as their implementation type and also a more generalised type.

Eg: Register `ConcreteService` as `IMyService`.

## Service lifetimes
There are two lifetimes in play: Per scenario and per test run.

## Services which are owned by the resolver
A service which is owned by the resolver is instantiated by the resolver via a factory, this it may be disposed by the resolver also.

If it was passed in as an instance then the resolver must not dispose it.

## Lazy resolution
The resolver must be able to provide lazy instances. They are instantiated (in their scope) on first use.

## Registry of services
It must be possible to inspect the registrations.

## Optional resolution
This means resolve s thing, but return null if it wasn't registered. Don't crash.
